dissect/target/plugins/os/unix/_os.py

```diff
@@ -37,27 +37,114 @@ def create(cls, target: Target, sysvol: Filesystem) -> UnixPlugin:
-                home=pwent[5],
-                shell=pwent[6],
-                _target=self.target,
-            )
+    @arg("--sessions", action="store_true", help="Parse syslog for recent user sessions")
+    def users(self, sessions: bool = False) -> Iterator[UnixUserRecord]:
+        """Recover users from /etc/passwd, /etc/master.passwd or /var/log/syslog session logins."""
+
+        # Yield users found in passwd files.
+        PASSWD_FILES = ["/etc/passwd", "/etc/master.passwd"]
+        USERS_CACHE = []
+
+        for passwd_file in PASSWD_FILES:
+            if (path := self.target.fs.path(passwd_file)).exists():
+                for line in path.open("rt"):
+                    line = line.strip()
+                    if not line or line.startswith("#"):
+                        continue
+
+                    pwent = dict(enumerate(line.split(":")))
+                    USERS_CACHE.append(pwent.get(0))
+                    yield UnixUserRecord(
+                        name=pwent.get(0),
+                        passwd=pwent.get(1),
+                        uid=pwent.get(2),
+                        gid=pwent.get(3),
+                        home=pwent.get(5),
+                        shell=pwent.get(6),
+                        source=passwd_file,
+                        _target=self.target,
+                    )
+
+        # Find users not in passwd files by parsing recent
+        # syslog ldap, kerberos and x-session logins.
+        # Must be enabled using the --sessions flag
+        if sessions and (path := self.target.fs.path("/var/log/syslog")).exists():
+            busses = []
+            sessions = []
+            cur_session = -1
+            needles = {
+                ("setting HOME=", "home"),
+                ("setting SHELL=", "shell"),
+                ("setting USER=", "name"),
+            }
+
+            for line in path.open("rt"):
+
+                # Detect the beginning of a new session activation in the syslog
+                #
+                # dbus-update-activation-environment starts a new session with
+                # DBUS_SESSION_BUS_ADDRESS, XDG_RUNTIME_DIR and/or DISPLAY
+                # Using DBUS_SESSION_BUS_ADDRESS seems to work fine for now.
+                if "setting DBUS_SESSION_BUS_ADDRESS=" in line:
+                    bus = line.split("DBUS_SESSION_BUS_ADDRESS=")[1].strip()
+                    if bus in busses:
+                        continue
+                    busses.append(bus)
+                    cur_session += 1
+                    sessions.append({"bus": bus, "name": None, "home": None, "shell": None, "uid": None})
+
+                # Search the following lines for more information on the previousley detected session.
+                for n, k in needles:
+                    if n in line and "dbus-update-activation-environment" in line:
+                        sessions[cur_session][k] = line.split(n)[1].strip()
+
+            for user in sessions:
```

Comment:
It could be intended behavior to have multiple sessions per user, as session details of a user can change over time. But I agree it would be the most logical approach from an IR perspective. Added de-duplication in [9740a1b](https://github.com/fox-it/dissect.target/pull/126/commits/9740a1bd9dc1db4a5d0b864e691cd6e32cff01b6)

Reply:
What about making `seen_users` a set with tuples of `(name, home, shell)`? That way you could re-use the `seen_users` set, and simultaneously deduplicate the sessions, while still getting new records for the same user if their details changed.

Reply:
If you could reflect this in the unit test that would be awesome :) Copy paste the current log lines twice, once with the same information and once with a different home or shell.

Reply:
Added in [1b846ab](https://github.com/fox-it/dissect.target/pull/126/commits/1b846aba5b748f2c7b7d1ebe5ffa17802a518522)